### DISCUSSION TITLE
Unique state names issue

### DISCUSSION THREAD

**Discussion by iamjuing** [StateSmith/StateSmith#125] [Ideas]
Hello,

got a question regarding this: [https://github.com/StateSmith/StateSmith/wiki/$mod-prefix]

Is there actually anything, that prevents StateSmith to prefix substate names with ther parent state names?
I mean, is there a use case atl all, that would not fit such a "hardcoded" behavior?

My point is to just simplify the usage as from the diagram, it is clear that states with same names and different parents, are just different states. StateSmith knows that from the diagram as well, so why to demand the user to use additional customization to allow that?

Thanks.

**Comment by adamfk** [comment]
Great question!

The original StateSmith would automatically prefix substate names with their parent's name. It worked OK for small designs, but many of the designs that I worked on at the time were for large GUIs that had a lot of nesting for menus. It quickly grew out of control. The state event handler name wouldn't fit on a screen anymore.

The other issue was that some older compilers had a small max length for identifiers and some coding rules like MISRA C dictate that max identifier length should be 31 characters.

The other practical issue to consider is the state ID to string function. If we prefix every substate name with their parent's name, the strings could be very large and eat up significant resources on smaller micros. 

Perhaps we could make a setting so that users could opt into automatic prefixing?

**Comment by iamjuing** [comment]
Fair points.
As a trade-off, the solution could be just to generate the state name in for "parentState_childState" only for the duplicates, and only up to the different parent name.
I guess, this would produce the same result as using the $mod prefix (that the user needs to use if there are two states with the same name); however with no extra effort in the diagram.

**Comment by adamfk** [comment]
I remember I started down that path of generating short unique names, but I can't remember why I stopped. I think it was just more effort than I wanted to put in at the time.

Care to open a ticket with a proposal for generating short unique names? It would be nice if users didn't have to rely on prefix modding unless they want custom control.

**Comment by iamjuing** [comment]
I don't think we need short names. Handlers shall still be readable enough for easy debugging.
I am not sure whether projects, using an ancient compiler that only supports identifier up to ~30 characters, incorporating a code generator is really a real world scenario.
When it comes to MISRA or a linter, I never apply them to generated files, as they are not under my control, so it makes no sense. It's a responsibility of supplier.
An acceptable way (at least for me) would be just print a warning that there are identifiers exceeding some (possibly configurable) limit and it is up to the designer to either accept it or rename the states so they are unique in the FSM.

**Comment by iamjuing** [comment]
Btw. the warning might be useful anyway, as the limit can be easily reached by the designer himself using long names, as the function names consist of ... stateName_signalName().

**Comment by adamfk** [comment]
Added an issue here: https://github.com/StateSmith/StateSmith/issues/137 .
